File: DSA05013.cpp
#include <iostream>
#include <vector>
using namespace std;
typedef vector<long long> vll;

const int mod = 1e9 + 7;
long long stairs (long long n, long long k) {
    vll dp(n + 1, 0);
    long long x = 1;
    long long sum = 0;
    for (int i = 1; i <= k; i ++) {
        dp[i] = x;
        x *= 2;
        x %= mod;
        sum += dp[i];
        sum %= mod;
    }
    for (int i = k + 1; i <= n ; i++) {
        sum %= mod;
        dp[i] = sum;
        sum -= dp[i - k];
        sum += dp[i];
    }
    return dp[n];
        
}

int main() {
    int t; cin >> t;
    while (t--) {
        int n, k;
        cin >> n >> k;
        cout << stairs (n , k)<< endl;
    }
}

File: DSA05013.py
from array import *


def function(n: int, k: int) -> int:
    mod = 1000000007
    if n <= k:
        return 2**(n - 1) % mod

    dp = [0 for i in range(n + 1)]

    sum = 0
    for i in range(1, k + 1):
        dp[i] = 2**(i - 1) % mod
        dp[i] %= mod
        sum += dp[i]
        sum %= mod

    for i in range(k + 1, n + 1):
        dp[i] = sum
        sum -= dp[i - k]
        sum += dp[i]
        sum %= mod

    return dp[n]


def main():
    for i in range(int(input())):
        n, k = map(int, input().split())
        print(function(n, k))


if __name__ == "__main__":
    main()
